### PR TITLE
Code object

### PR DESCRIPTION
**Pull request by cthulahoops**
Still work in progress (but would be okay to merge).

### CODE REVIEW

**Comment by cthulahoops** [comment]
This should be the uncontroversial stuff. All these concepts exist and are well supported in RustPython, and I think this is ready to merge. (CI permitting)

The next steps are a bit tougher:

We have freevars, locals, cellvars, etc. none of which exist as concepts yet. We probably want them for efficiency eventually.

Then we have co_code which returns the bytecode encoded as a binary which is then interpreted using ```opcode.opname```.

There are two approaches here:

* Write an encoder / decoder to serialise the bytecode to a binary, and generate the maps in opcode to decode them. Once everything is implemented the ```dis``` module should just work.
* Re-implement the dis module in rust using the native types provided by bytecode.

I'm still leaning towards the former, but I'm not sure.

File: tests/snippets/code.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+c1 = compile("1 + 1", "", 'eval')
+
+code_class = type(c1)
```
Comment (windelbouwman) [inline tests/snippets/code.py]: We should have `types.CodeType`.

**Comment by windelbouwman** [comment]
Do we require freevars, locals and cellvars? I am unaware what they are. If they are CPython implementation detail, I suggest to leave it out. If it is essential Python behavior, we must have them.

I think it would be a good step to have a code object serializer / deserializer. This also enables us to cache pyc files on disk, like CPython does. I think this would be a good next step? Then the dis module becomes indeed easy, as you mentioned.

**Comment by cthulahoops** [comment]
I can't tell whether they are a cpython detail, or something we have to implement. Is the code class api part of the language? It's certainly not well documented. On the other hand, I think we will eventually need them for speed reasons. (Locals, for example, is used by the FAST_LOAD and FAST_STORE operations to make local variables a simple array operation rather than a python dictionary operation.)

Cool, I'll do the bytecode serialisation/deserialisation next.

**Pull request merged**